contracts/btc-finality/src/finality.rs

```diff
@@ -891,52 +902,61 @@ pub fn query_fps_by_total_active_sats(
+pub fn update_rewards_dist_for_interval(
+    deps: &mut DepsMut,
+    env: &Env,
+) -> Result<(), ContractError> {
     let cfg = CONFIG.load(deps.storage)?;
-    let rewards_amount = deps
+
+    // Get current balance of the finality contract (total rewards to distribute)
+    let current_balance = deps
         .querier
         .query_balance(env.contract.address.clone(), cfg.denom)?
-        .amount
-        .saturating_sub(total_pending_rewards);
-    // Short-circuit if there are no rewards to distribute
-    if rewards_amount.is_zero() {
-        return Ok(());
+        .amount;
+
+    if current_balance.is_zero() {
+        return Ok(()); // No rewards to distribute
     }
-    // Compute the rewards for each active FP
-    let mut accumulated_rewards = Uint128::zero();
-    for (fp_btc_pk_hex, power) in active_fps {
-        let reward = (rewards_amount.u128() * power as u128) / total_voting_power;
-        let reward = Uint128::from(reward);
-        // Update the rewards for this FP
-        REWARDS.update(deps.storage, &fp_btc_pk_hex, |r| {
-            Ok::<Uint128, ContractError>(r.unwrap_or_default() + reward)
-        })?;
-        // Compute the total rewards
-        accumulated_rewards += reward;
+
+    // Single iteration: calculate total weight and distribute rewards
+    let mut total_accumulated_weight = 0u128;
+    let mut fp_entries = Vec::new();
+
+    // First pass: collect entries and calculate total
+    for item in
+        ACCUMULATED_VOTING_WEIGHTS.range(deps.storage, None, None, cosmwasm_std::Order::Ascending)
+    {
+        let (fp_btc_pk_hex, weight) = item?;
+        total_accumulated_weight += weight;
+        fp_entries.push((fp_btc_pk_hex, weight));
     }
-    // Update the total rewards
-    TOTAL_PENDING_REWARDS.update(deps.storage, |r| {
-        Ok::<Uint128, ContractError>(r + accumulated_rewards)
-    })?;
+
+    if fp_entries.is_empty() || total_accumulated_weight == 0 {
+        return Ok(()); // No accumulated voting weights or zero total
+    }
+
+    // Convert total_accumulated_weight to Uint128 for safer arithmetic
+    let total_accumulated_weight = Uint128::from(total_accumulated_weight);
+
+    // Distribute rewards proportionally based on accumulated voting weights
+    for (fp_btc_pk_hex, accumulated_weight) in fp_entries {
+        // Use Uint128 arithmetic for safe multiplication and division with floor division
+        let accumulated_weight = Uint128::from(accumulated_weight);
+        let numerator = current_balance.checked_mul(accumulated_weight)?;
+        let reward = numerator.div_floor((total_accumulated_weight, Uint128::one()));
+
+        if !reward.is_zero() {
+            // Add reward to FP's pending rewards
+            REWARDS.update(deps.storage, &fp_btc_pk_hex, |existing| {
+                Ok::<Uint128, ContractError>(existing.unwrap_or_default() + reward)
+            })?;
+        }
```

Comment:
This `if` seems not necessary

Reply:
I feel it's hard to reason `reward` can never be zero, so keeping it just in case